### ISSUE TITLE
Individual injector to each pattern?

### ISSUE THREAD

**Issue by 7iomka** [yenshih/style-resources-loader#5]
Hello! Plugin is awesome! One question: It's possible to make works individual injector to each pattern?
I have an components library, and i want to make this works: http://s3.micp.ru/BPVBj.jpg
In this short example file named as btn.scss don't know variables defined in btn.config.scss, same as file btn.override.scss don't know map variables, defined in btn.config.scss;

I don't want to do internal imports in scss files, because this slows down the rebuild of the project, and also makes it impossible to import a new file without access to the file itself. I want to make sure that every component in the components folder, for example btn, is recognized by the loader by the mask. I think that this idea would make it work similarly to imports in scss.

Thanks for any reply!

**Comment by yenshih** [comment]
Hi! May I have more specific examples about your component library? Such as those `scss` files content and expected behaviors.
IMO, individual injector to each pattern is supported.
```js
patterns: [
    '*.override.scss',
    '*.config.scss',
    '*.scss'
],
injector: (source, resources) => {
    const configResources = resources.filter(({ file }) => file.endsWith('config.scss'));
    const overrideResources = resources.filter(({ file }) => file.endsWith('override.scss'));
    // ...
}
```
Hope it helps!

**Comment by 7iomka** [comment]
Ok. In my app I have common.js entry:

> common.js
```
// base styles, reset.css, functions, mixins, core config and etc..
import '../scss/app.scss'; 

// button component
import '../components/btn/btn';
//... other components
```

In this example button component don't know about variables and some global mixins, imported in app.scss. For this case your plugin I use like this:
```

{
    loader: 'style-resources-loader',
    options: {
        patterns: [
            /*=============================================
            =            Core functions            =
            =============================================*/
            path.resolve(__dirname, '../dev/scss/core/_functions.scss'),

            /*=====  End of Core functions  ======*/


            /*=============================================
            =            Color palette variables            =
            =============================================*/
            path.resolve(__dirname, '../dev/scss/core/_palette.scss'),
            /*=====  End of Color palette variables  ======*/


            /*=============================================
            =            Global variables settings           =
            =============================================*/
            path.resolve(__dirname, '../dev/scss/core/_settings.scss'),

            /*=====  End of Global project settings  ======*/


            /*=============================================
            =            Mixins for common (global) usage            =
            =============================================*/
            path.resolve(__dirname, '../dev/scss/core/_mixins.scss'),

            /*=====  End of Mixins for common (global) usage  ======*/


            /*=============================================
            =            Breackpoints for media queries            =
            =============================================*/
            path.resolve(__dirname, '../dev/scss/core/_media.scss'),
            path.resolve(__dirname, '../dev/scss/core/helpers/_include-media.scss'),
        ],
        injector: 'prepend', // mb: append | func
    }
},

```

This config make possible to read in components sass files global variables / mixins / functions;
All woks fine.

But now let's look at the button component:
I have a folder named btn, which contains files:
> btn.js
> btn.scss
> btn.config.scss
> btn.override.scss

The task is to import, in addition to the styles in the btn.js component file, the configuration files (which contain a sass map with default settings (btn.config.scss), and a map with custom user settings (btn.override.scss  - that overwrite those that are set by default);
Currently I'm doing this inside a stylesheet file (btn.scss):
> btn.scss
```
@import './btn.override.scss';
@import './btn.config.scss';

// some styles that uses config variables from sass resolved map config
```
In this model of import in sass files, there are drawbacks:
1) Files imported in sass files (nested imports) works very slow in incremental build mode (i tested this by making changes in separate component independed file like button, in comparation with file like bootstrap.scss, which use multiple internal imports)
2) The user should not modify the component file, however, if he needs to connect another file, for example btn.custom.scss, which need to have acess to config variables without internal imports in sass, he has no options to do it!

And I got the idea!
What I want?
In each scss file, imported in general js file of component, (in our case > btn.js) must be imported (if exist) in strict order user overrides and default config sass files. In our example we have only one single import in btn.js
> btn.js
```
import './btn.scss'; 

```
Loader see this import rule, and make this before import it entry:
```
@import './btn.override.scss';
@import './btn.config.scss';
```

He needs to do this for each component individually, where the name of the imported file will be a variable `${name_of_component}.override.scss`, `${name_of_component}.config.scss`;
In addition, is not necessarily, but it is desirable that the plugin check the existence of such files in the component folder, because some components I have are simple enough to have a configuration file.

Thank you for your ideas and for plugin!

**Comment by yenshih** [comment]
Thank you for your explanation. I'm trying to understand what you have said.
Assume that we have a `components` folder like this:
```
/components
    |-- btn.js
    |-- btn.scss
    |-- btn.override.scss
    |-- btn.config.scss
    |-- input.js
    |-- input.scss
    |-- input.config.scss
```
What you want is to replace those `${component}.override.scss` and `${component}.config.scss` `@import` statements in each `${component}.scss` file, since you found internal imports would cause a performance issue.
I guess you need individual injector to each **source/entry file** (such as `btn.scss`, `input.scss`), not patterns. Am I right?

Unfortunately, we have no way to get source file name or path in loader. See [#6124](https://github.com/webpack/webpack/issues/6124).

Indeed, we do have multiple workarounds:
 - Optimize your build performance, see [happypack](https://github.com/amireh/happypack).
 - Hardcode some metadata in your source `scss` file (such as `// @file btn.scss`), write your own injector and parse those metadata. Finally, you can get component name and do what you want. It's tricky.
 - Use multiple loader modules to resolve each component style. However, this may cause worse performance issue.

**Comment by 7iomka** [comment]
> I guess you need individual injector to each source/entry file (such as btn.scss, input.scss), not patterns. Am I right?

Yes! I really did not know that we do not have access to source file name or path in loader :(
Happypack works fine, only in my build it does not fit due to the specificity of the settings.
Hardcoded metadata it's worse than imports in sass))
I wanted to avoid imports into the sass files, at least in order to maintain a single format of imports in js files (all dependencies expect corresponded import in js file of component).

It turns out that at the moment, there is nothing better than to put up with the situation and use the import in styles in different files every time when we need component configurations?
In any case, thanks for listening!)

**Comment by yenshih** [comment]
> It turns out that at the moment, there is nothing better than to put up with the situation and use the import in styles in different files every time when we need component configurations?

I'm afraid so, unless you use "multiple configurations" (quote from [#6124](https://github.com/webpack/webpack/issues/6124)).

Maybe `@import` in `scss` is fine :)
